In [1]:
# account credentials
username = "krishnaraj.kpt@outlook.com"
password = "BBQtJTSs8uQh57aw"

imap_server = "outlook.office365.com"

In [8]:
import imaplib
import email
from email.header import decode_header
import json


In [15]:
# create an IMAP4 class with SSL
imap = imaplib.IMAP4_SSL(imap_server)
# authenticate
imap.login(username, password)
status, messages = imap.select("INBOX")

# number of top emails to fetch
N = 850

# total number of emails
messages = int(messages[0])
print(messages)

1011


In [7]:
# get all header information for each message


# fetch the top N email headers
for i in range(N):
    # fetch the email header
    result, data = imap.fetch(str(i + 1), "(RFC822.HEADER)")
    if result == "OK":
        # parse the email header
        email_header = email.message_from_bytes(data[0][1])

        # print all the available headers
        for header in email_header.items():
            print(f"{header[0]}: {header[1]}")

# close the IMAP connection
imap.close()

MIME-Version: 1.0
Received: from CO1NAM05HT196.eop-nam05.prod.protection.outlook.com
 (2603:10b6:408:94::41) by BN7PR07MB4739.namprd07.prod.outlook.com with HTTPS
 via BN8PR03CA0028.NAMPRD03.PROD.OUTLOOK.COM; Mon, 27 Jan 2020 15:47:45 +0000
Received: from CO1NAM05FT061.eop-nam05.prod.protection.outlook.com
 (2a01:111:e400:7e50::34) by
 CO1NAM05HT196.eop-nam05.prod.protection.outlook.com (2a01:111:e400:7e50::398)
 with Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.2686.12; Mon, 27 Jan
 2020 15:47:44 +0000
Authentication-Results: spf=pass (sender IP is 136.147.186.38)
 smtp.mailfrom=bounce.microsoftemails.com; outlook.com; dkim=pass (signature
 was verified) header.d=mail.onedrive.com;outlook.com; dmarc=pass action=none
 header.from=mail.onedrive.com;
Received-SPF: Pass (protection.outlook.com: domain of
 bounce.microsoftemails.com designates 136.147.186.38 as permitted sender)
 receiver=protection.outlook.com; client-ip=136.147.186.38;
 h

KeyboardInterrupt: 

In [4]:
for i in range(messages, messages-N, -1):
    # fetch the email message by ID
    res, msg = imap.fetch(str(i), "(RFC822)")
    print(msg)
    for response in msg:
        try:
            if isinstance(response, tuple):
                # parse a bytes email into a message object
                msg = email.message_from_bytes(response[1])
                # decode the email subject
                headers = decode_header( msg["Subject"] )
                print(headers)
        except Exception as e:
            print(e)
            pass
# close the connection and logout
imap.close()
imap.logout()

1003
[(b'1003 (RFC822 {49883}', b'Received: from DM8PR17MB4952.namprd17.prod.outlook.com (2603:10b6:8:3e::13) by\r\n DS7PR17MB6730.namprd17.prod.outlook.com with HTTPS; Sat, 28 Oct 2023 07:13:37\r\n +0000\r\nReceived: from DUZPR01CA0082.eurprd01.prod.exchangelabs.com\r\n (2603:10a6:10:46a::10) by DM8PR17MB4952.namprd17.prod.outlook.com\r\n (2603:10b6:8:3e::13) with Microsoft SMTP Server (version=TLS1_2,\r\n cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.6933.14; Sat, 28 Oct\r\n 2023 07:13:35 +0000\r\nReceived: from DB8EUR06FT057.eop-eur06.prod.protection.outlook.com\r\n (2603:10a6:10:46a:cafe::9c) by DUZPR01CA0082.outlook.office365.com\r\n (2603:10a6:10:46a::10) with Microsoft SMTP Server (version=TLS1_2,\r\n cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.6933.24 via Frontend\r\n Transport; Sat, 28 Oct 2023 07:13:35 +0000\r\nAuthentication-Results: spf=pass (sender IP is 192.254.112.98)\r\n smtp.mailfrom=sgmail.github.com; dkim=pass (signature was verified)\r\n header.d

KeyboardInterrupt: 

In [10]:

# create a list to store the emails
emails = []

# fetch the top N email headers
for i in range(N):
    # fetch the email header
    result, data = imap.fetch(str(i + 1), "(RFC822.HEADER)")
    if result == "OK":
        # parse the email header
        email_header = email.message_from_bytes(data[0][1])

        # create a dictionary to store the email header
        email = {
            "subject": email_header["Subject"],
            "headers": [],
        }

        # add the email headers to the dictionary
        for header in email_header.items():
            email["headers"].append({
                "header": header[0],
                "value": header[1],
            })

        # add the email to the list of emails
        emails.append(email)

# close the IMAP connection
imap.close()

# write the emails to a JSON file
with open("emails.json", "w") as f:
    json.dump(emails, f, indent=4)

AttributeError: 'dict' object has no attribute 'message_from_bytes'

In [19]:
import imaplib
import email
from email.parser import BytesParser
from email.header import decode_header
import json

# create an IMAP4 class with SSL
imap = imaplib.IMAP4_SSL(imap_server)

# authenticate
imap.login(username, password)

# select the INBOX folder
status, messages = imap.select("INBOX")

# number of top emails to fetch
N = 1014

# total number of emails
messages = int(messages[0])

# create a list to store the emails
emails = []

# fetch the top N email headers
for i in range(1000, N):
    # fetch the email header
    result, data = imap.fetch(str(i + 1), "(RFC822.HEADER)")
    if result == "OK":
        # parse the email header
        email_parser = BytesParser()
        email_header = email_parser.parsebytes(data[0][1])

        # create a dictionary to store the email header
        email = {
            "subject": email_header["Subject"],
            "headers": [],
        }

        # add the email headers to the dictionary
        for header in email_header.items():
            email["headers"].append({
                "header": header[0],
                "value": header[1],
            })

        # add the email to the list of emails
        emails.append(email)

# close the IMAP connection
imap.close()

# write the emails to a JSON file
with open("emails.json", "w") as f:
    json.dump(emails, f, indent=4)

Certainly, here's a point-by-point summary of the email's journey as inferred from the provided email headers:

Sender and Origin:

The email was sent by "Parth Zarekar" from the email address "parthzarekar@gmail.com."
The email's subject is "Sending from browser using mobile hotspot."
Received from Outlook.com:

The email was first received from "SJ0PR17MB4597.namprd17.prod.outlook.com" with an IPv6 address (2603:10b6:a03:35a::11).
It was delivered to "DS7PR17MB6730.namprd17.prod.outlook.com" via HTTPS.
Intermediate Servers:

The email passed through "AS4P190CA0019.EURP190.PROD.OUTLOOK.COM" and "AM7EUR06FT061.eop-eur06.prod.protection.outlook.com."
Each server is identified with IPv6 addresses.
Authentication Checks:

The email passed Sender Policy Framework (SPF), DomainKeys Identified Mail (DKIM), and Domain-based Message Authentication, Reporting, and Conformance (DMARC) checks, indicating its authenticity.
SPF passed, confirming that the sender's IP (209.85.208.41) is authorized to send on behalf of gmail.com.
DKIM passed, validating the email's digital signature.
DMARC passed with an action set to "none," suggesting a non-strict policy.
X-Received Header:

The email's journey included Google's servers, as indicated by "mail-ed1-f41.google.com."
This header provides a timestamp for when the email reached its destination.
Message-ID:

The email has a unique identifier, "CAAVEJ5Tt1R4nJsCPE_UcVe0Hb1c8n4VY3FC6VefzhzXFe-Mg@mail.gmail.com," which can be used for tracking.
    Recipient:

The recipient's email address is "krishnaraj.kpt@outlook.com."
Content-Type Header:

The email is of "multipart/alternative" type, which suggests it may contain both plain text and HTML versions.
X-MS-Exchange Headers:

These headers pertain to Microsoft Exchange's processing of the email.
X-Microsoft-Antispam Headers:

These headers contain information related to Microsoft's antispam measures and classification.
X-Message-Info Header:

This header appears to be an identifier or marker for the email.
    X-Message-Delivery Header:

Provides information about the delivery of the email.
In summary, the email originated from "Parth Zarekar" and passed through Outlook.com, Google's servers, and Microsoft Exchange infrastructure. It successfully passed authentication checks, ensuring its legitimacy. The recipient of the email is "krishnaraj.kpt@outlook.com." The headers also include various details related to email processing, antispam measures, and unique identifiers for tracking.

Analyzing email headers can provide a wealth of information about the email's origin, delivery path, and security features. Here's a thorough analysis of the email headers you provided:

Subject: The subject of the email is "Sending from browser using mobile hotspot."

Received Headers:

These headers show the path the email took from the sender to the recipient's inbox.
The email originated from an Outlook.com server, was sent over HTTPS, and went through several intermediate servers.
It includes information about server versions, encryption methods, and timestamps.
Authentication-Results Header:

This header indicates that the email passed Sender Policy Framework (SPF), DomainKeys Identified Mail (DKIM), and Domain-based Message Authentication, Reporting, and Conformance (DMARC) checks.
SPF passed, indicating that the sender's IP address (209.85.208.41) is authorized to send on behalf of gmail.com.
DKIM passed, verifying that the email's digital signature is valid.
DMARC also passed with the action set to "none," indicating that the DMARC policy is not strict.
Received-SPF Header:

SPF authentication passed, with the domain gmail.com designating 209.85.208.41 as a permitted sender.
DKIM-Signature Header:

This header contains information related to DomainKeys Identified Mail (DKIM) signature.
It includes details about the encryption and signing methods used to verify the email's authenticity.
X-Google-DKIM-Signature Header:

Similar to the DKIM-Signature header, this one is specific to Google's DKIM signature.
X-Gm-Message-State Header:

This header appears to be related to Google's internal message state tracking.
X-Google-Smtp-Source Header:

Provides information about the source of the email as it relates to Google's SMTP service.
X-Received Header:

Indicates that the email was received by an IPv6 address and provides a timestamp.
From Header:

Specifies the sender's name and email address as "Parth Zarekar parthzarekar@gmail.com."
Date Header:

Indicates the date and time when the email was sent, which is "Mon, 30 Oct 2023 00:02:20 +0530."
Message-ID Header:

Provides a unique identifier for the email, such as "CAAVEjOJ5Tt1R4nJsCPE_UcVe0Hb1c8n4VY3FC6VefzhzXFe-Mg@mail.gmail.com."
To Header:

Specifies the recipient's email address as "krishnaraj.kpt@outlook.com."
Content-Type Header:

Indicates that the email contains multipart/alternative content with a specific boundary.
X-MS-Exchange-Organization-ExpirationStartTime Header:

Contains information related to the email's expiration time within the Microsoft Exchange organization.
X-MS-Exchange-Organization-ExpirationInterval Header:

Indicates the interval for which the email is set to expire.
X-MS-Exchange-Organization-Network-Message-Id Header:

Provides a unique network message ID.
X-MS-Exchange-Organization-AuthSource and X-MS-Exchange-Organization-AuthAs Headers:

These headers specify the source and authentication status of the email.
X-Sender-IP Header:

Displays the sender's IP address as "209.85.208.41."
X-SID-PRA Header:

Indicates the "PRA" (Purported Responsible Address) for the email.
    X-SID-Result Header:

Shows that the email passed authentication with a "PASS" result.
X-MS-Exchange-Organization-PCL Header:

Specifies the PCL (Precedence, which is a spam classification) for the email.
    X-Microsoft-Antispam Header:

Contains information related to Microsoft's antispam classification.
X-MS-Exchange-CrossTenant Headers:

These headers are related to the email's path through Microsoft's Exchange infrastructure.
X-MS-Exchange-Transport-EndToEndLatency Header:

Indicates the latency experienced during email transport.
X-MS-Exchange-Processed-By-BccFoldering Header:

Specifies the processing information for the email.
    X-Microsoft-Antispam-Mailbox-Delivery Header:

Contains information about antispam measures and email delivery.
X-Message-Info Header:

Appears to be an identifier or marker for the email.
    X-Message-Delivery Header:

Provides information about the email delivery.
X-Microsoft-Antispam-Message-Info Header:

Contains additional information related to Microsoft's antispam measures.
These email headers provide detailed insights into the email's journey, authentication checks, and various characteristics related to its transmission and processing.

# Checking the location of an ip address

In [23]:
import requests

def get_ip_location(ip_address):
    # Make a GET request to ipinfo.io with the IP address
    url = f"https://ipinfo.io/{ip_address}/json"
    response = requests.get(url)

    if response.status_code == 200:
        data = response.json()
        return data
    else:
        return None

# Input the IPv4 address for which you want to retrieve the location
ip_address = "2402:e280:3e28:3a0:3c6f:b8b9:1cb6:2265"  # Replace with the IP address you want to look up

location_info = get_ip_location(ip_address)
print(location_info)
if location_info:
    # Print the location information
    print(f"IP Address: {location_info['ip']}")
    print(f"Hostname: {location_info['hostname']}")
    print(f"City: {location_info['city']}")
    print(f"Region: {location_info['region']}")
    print(f"Country: {location_info['country']}")
    print(f"Location: {location_info['loc']}")
    print(f"Organization: {location_info['org']}")
    print(f"Timezone: {location_info['timezone']}")
else:
    print("Unable to retrieve location information for the IP address.")


{'ip': '2402:e280:3e28:3a0:3c6f:b8b9:1cb6:2265', 'city': 'Pune', 'region': 'Maharashtra', 'country': 'IN', 'loc': '18.5196,73.8554', 'org': 'AS134674 TATA PLAY BROADBAND PRIVATE LIMITED', 'postal': '411001', 'timezone': 'Asia/Kolkata', 'readme': 'https://ipinfo.io/missingauth'}
IP Address: 2402:e280:3e28:3a0:3c6f:b8b9:1cb6:2265


KeyError: 'hostname'